internal/envoy/versions.go

```diff
@@ -20,30 +20,94 @@ import (
 	"fmt"
 	"io"
 	"net/http"
+	"strconv"
+	"strings"
 
+	"github.com/tetratelabs/func-e/internal/globals"
 	"github.com/tetratelabs/func-e/internal/version"
 )
 
-// FuncEVersions returns a version map from a remote URL. eg globals.DefaultEnvoyVersionsURL.
-func FuncEVersions(ctx context.Context, envoyVersionsURL string, p version.Platform, v version.Version) (version.ReleaseVersions, error) {
+// NewFuncEVersions creates a new Envoy versions fetcher.
+func NewFuncEVersions(envoyVersionsURL string, p version.Platform, v version.Version) version.FuncEVersions {
+	feV := &funcEVersions{envoyVersionsURL: envoyVersionsURL, platform: p, version: v}
+	feV.getFunc = feV.Get
+	return feV
+}
+
+type funcEVersions struct {
+	envoyVersionsURL string
+	platform         version.Platform
+	version          version.Version
+
+	// getFunc allows to override the release versions getter implementation.
+	getFunc func(ctx context.Context) (version.ReleaseVersions, error)
+}
+
+var _ version.FuncEVersions = (*funcEVersions)(nil)
+
+// Get implements fetching the Envoy versions from the specified Envoy versions URL.
+func (f *funcEVersions) Get(ctx context.Context) (version.ReleaseVersions, error) {
 	result := version.ReleaseVersions{}
 	// #nosec => This is by design, users can call out to wherever they like!
-	resp, err := httpGet(ctx, envoyVersionsURL, p, v)
+	resp, err := httpGet(ctx, f.envoyVersionsURL, f.platform, f.version)
 	if err != nil {
 		return result, err
 	}
 	defer resp.Body.Close() //nolint
 
 	if resp.StatusCode != http.StatusOK {
-		return result, fmt.Errorf("received %v status code from %v", resp.StatusCode, envoyVersionsURL)
+		return result, fmt.Errorf("received %v status code from %v", resp.StatusCode, f.envoyVersionsURL)
 	}
 	body, err := io.ReadAll(resp.Body) // fully read the response
 	if err != nil {
-		return result, fmt.Errorf("error reading %s: %w", envoyVersionsURL, err)
+		return result, fmt.Errorf("error reading %s: %w", f.envoyVersionsURL, err)
 	}
 
 	if err := json.Unmarshal(body, &result); err != nil {
 		return result, fmt.Errorf("error unmarshalling Envoy versions: %w", err)
 	}
 	return result, nil
 }
+
+// FindLatestPatch implements finding the latest patch version for the given minor version or raises
+// an error. The Envoy release versions fetching logic can be overridden by setting the getFunc with
+// different implementation.
+func (f *funcEVersions) FindLatestPatch(ctx context.Context, minorVersion version.Version) (version.Version, error) {
```

Comment:
This will be used later when we introduce `func-e use <minor version>`, as an example: `func-e use 1.18`